release-19.1: storage: create new TestSnapshotAfterTruncationWithUncommittedTail test

nvb opened the pull request:
Backport:
  * 2/2 commits from "storage: create new TestSnapshotAfterTruncationWithUncommittedTail test" (#37058)
  * 1/1 commits from "storage: speed up TestSnapshotAfterTruncationWithUncommittedTail" (#37079)

Please see individual PRs for details.

/cc @cockroachdb/release

nvb committed:
storage: filter Raft heartbeats in unreliableRaftHandler

Release note: None

nvb committed:
storage: create new TestSnapshotAfterTruncationWithUncommittedTail test

This PR adds a regression test for #37056. In doing so, it also confirms
the theory that #37055 is the proper fix for that bug.

Before #37055, the test would get stuck with the following logs repeatedly
firing:
```
I190424 00:15:52.338808 12 storage/client_test.go:1242  SucceedsSoon: expected [21 21 21], got [12 21 21]
I190424 00:15:52.378060 78 vendor/go.etcd.io/etcd/raft/raft.go:1317  [s1,r1/1:/M{in-ax}] 1 [logterm: 6, index: 31] rejected msgApp [logterm: 8, index: 31] from 2
I190424 00:15:52.378248 184 vendor/go.etcd.io/etcd/raft/raft.go:1065  [s2,r1/2:/M{in-ax}] 2 received msgApp rejection(lastindex: 31) from 1 for index 31
I190424 00:15:52.378275 184 vendor/go.etcd.io/etcd/raft/raft.go:1068  [s2,r1/2:/M{in-ax}] 2 decreased progress of 1 to [next = 31, match = 31, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
```

After #37055, the test passes.

Release note: None

nvb committed:
storage: speed up TestSnapshotAfterTruncationWithUncommittedTail

This commit speeds up the test added in #37058 by parallelizing its
failed writes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37093)
<!-- Reviewable:end -->

nvb committed:
storage: use DistSender when possible in TestSnapshotAfterTruncationWithUncommittedTail

This avoids unnecessary NotLeaseHolderErrors when we don't need to enforce
which replica can evaluate a request. We see that in the current CI failure
in #37093.

Release note: None

nvb committed:
storage: fix incorrect Raft message filtering, which made test flaky

Fixes #37085.

This filtering was assuming that the Raft message's index field
indicated the first index in the MsgApp. It actually indicates the
log index that _precedes_ any of the entries in the MsgApp.

It stole this from TestReplicaRangefeedRetryErrors, so the commit
updates that too.

Release note: None
